src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,337 @@
+from flint.flintlib.flint cimport slong
+from flint.flintlib.fmpz cimport (
+    fmpz_t,
+    fmpz_one,
+    fmpz_set,
+    COEFF_IS_MPZ,
+    fmpz_get_str,
+    fmpz_init,
+    fmpz_clear
+)
+from flint.flintlib.fmpz_mod cimport *
+
+from flint.utils.typecheck cimport typecheck
+from flint.utils.conversion cimport str_from_chars
+from flint.flint_base.flint_base cimport flint_scalar
+from flint.types.fmpz cimport (
+    fmpz,
+    any_as_fmpz,
+)
+
+cimport libc.stdlib
+
+# TODO: import this from types.fmpz somehow, it's not good
+# to have the function repeated here.
+cdef fmpz_get_intlong(fmpz_t x):
+    """
+    Convert fmpz_t to a Python int or long.
+    """
+    cdef char * s
+    if COEFF_IS_MPZ(x[0]):
+        s = fmpz_get_str(NULL, 16, x)
+        v = int(str_from_chars(s), 16)
+        libc.stdlib.free(s)
+        return v
+    else:
+        return <slong>x[0]
+
+cdef class fmpz_mod_ctx:
+    """
+    """
+    def __cinit__(self, mod):
+        self.__init__(mod)
+
+    def __dealloc__(self):
+        fmpz_mod_ctx_clear(self.val)
+
+    def __init__(self, mod):
+        """
+        """
+        # Ensure modulus is fmpz type
+        if not typecheck(mod, fmpz):
+            mod = any_as_fmpz(mod)
+            if mod is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+
+    def __eq__(self, other):
+        if typecheck(other, fmpz_mod_ctx):
+            return self.val.n == (<fmpz_mod_ctx>other).val.n
+        return False
```

Comment:
Maybe it should be `self.modulus() == other.modulus()`.

Here you are directly accessing `self.val.n` which is an `fmpz_t`. Those cannot be compared directly at the C level (they might be pointers): you would need to use `fmpz_equal` to compare them in C.

Reply:
I need to be careful actually that the same modulus but different context object doesn't cause issues.

Reply:
There could be a global context cache using `WeakValueDictionary`:
https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary

```python
from weakref import WeakValueDictionary
cache = WeakValueDictionary()

def get_context(n):
    ctx = cache.get(n)
    if ctx is not None:
        return ctx
    ctx = cache.setdefault(n, fmpz_mod_ctx(n))
    return ctx
```
https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#making-extension-types-weak-referenceable

Then there would only ever be one context object for a given modulus in memory at a time.

The advantage of using `WeakValueDictionary` rather than an ordinary `dict` is that the context will be freed when it is not needed any more.